alf/algorithms/dynamics_learning_algorithm.py

```diff
@@ -0,0 +1,226 @@
+
+
+@gin.configurable
+class DeterministicDynamicsAlgorithm(DynamicsLearningAlgorithm):
+    """Deterministic Dynamics Learning Module
+
+    This module trys to learn the dynamics of environment with a
+    determinstic model.
+    """
+
+    def __init__(self,
+                 action_spec,
+                 feature_spec,
+                 hidden_size=256,
+                 dynamics_network: Network = None,
+                 name="DeterministicDynamicsAlgorithm"):
+        """Create a DeterministicDynamicsAlgorithm.
+
+        Args:
+            hidden_size (int|tuple): size of hidden layer(s)
+            dynamics_network (Network): network for predicting next feature
+                based on the previous feature and action. It should accept
+                input with spec [feature_spec, encoded_action_spec] and output
+                a tensor of shape feature_spec. For discrete action,
+                encoded_action is an one-hot representation of the action.
+                For continuous action, encoded action is the original action.
+        """
+        if dynamics_network is not None:
+            dynamics_network_state_spec = dynamics_network.state_spec
+        else:
+            dynamics_network_state_spec = ()
+
+        super().__init__(
+            train_state_spec=DynamicsState(
+                feature=feature_spec, network=dynamics_network_state_spec),
+            action_spec=action_spec,
+            feature_spec=feature_spec,
+            hidden_size=hidden_size,
+            dynamics_network=dynamics_network,
+            name=name)
+
+    def predict(self, time_step: ActionTimeStep, state: DynamicsState):
+        """Predict the next observation given the current time_step.
+                The next step is predicted using the prev_action from time_step
+                and the feature from state.
+        """
+        action = self._encode_action(time_step.prev_action)
+        obs = state.feature
+        forward_delta, network_state = self._dynamics_network(
+            inputs=[obs, action], network_state=state.network)
+        forward_pred = obs + forward_delta
+        state = state._replace(feature=forward_pred, network=network_state)
+        return AlgorithmStep(outputs=forward_pred, state=state, info=())
+
+    def update_state(self, time_step: ActionTimeStep, state: DynamicsState):
+        """Update the state based on ActionTimeStep data. This function is
+            mainly used during rollout together with a planner
+        Args:
+            time_step (ActionTimeStep): input data for dynamics learning
+            state (Tensor): state for DML (previous observation)
+        Returns:
+            TrainStep:
+                outputs: empty tuple ()
+                state: feature
+                info (DynamicsInfo):
+        """
+        feature = time_step.observation
+        return state._replace(feature=feature)
+
+    def train_step(self, time_step: ActionTimeStep, state: DynamicsState):
```

Comment:
ActionTimeStep

Reply:
You have looked into the TF version (the first commit). This is only used for reference.